UefiPayloadPkg/Library/PciSegmentInfoLibAcpiBoardInfo/PciSegmentInfoLibAcpiBoardInfo.c

```diff
@@ -55,18 +57,53 @@ RetrieveMultiSegmentInfoFromHob (
   mPciSegments = (PCI_SEGMENT_INFO *)AllocatePool (Size);
   ASSERT (mPciSegments != NULL);
   ZeroMem (mPciSegments, PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO));
+  pPciSegments = (PCI_SEGMENT_INFO *)AllocatePool (Size);
+  ASSERT (pPciSegments != NULL);
+  ZeroMem (pPciSegments, PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO));
 
   //
   // Create all root bridges with PciRootBridgeInfoHob
   //
   for (Index = 0; Index < PciRootBridgeInfo->Count; Index++) {
     if (UplSegmentInfo->SegmentInfo[Index].SegmentNumber == (UINT16)(PciRootBridgeInfo->RootBridge[Index].Segment)) {
-      mPciSegments[Index].BaseAddress = UplSegmentInfo->SegmentInfo[Index].BaseAddress;
+      pPciSegments[Index].BaseAddress = UplSegmentInfo->SegmentInfo[Index].BaseAddress;
     }
 
-    mPciSegments[Index].SegmentNumber  = (UINT16)(PciRootBridgeInfo->RootBridge[Index].Segment);
-    mPciSegments[Index].StartBusNumber = (UINT8)PciRootBridgeInfo->RootBridge[Index].Bus.Base;
-    mPciSegments[Index].EndBusNumber   = (UINT8)PciRootBridgeInfo->RootBridge[Index].Bus.Limit;
+    pPciSegments[Index].SegmentNumber  = (UINT16)(PciRootBridgeInfo->RootBridge[Index].Segment);
+    pPciSegments[Index].StartBusNumber = (UINT8)PciRootBridgeInfo->RootBridge[Index].Bus.Base;
+    pPciSegments[Index].EndBusNumber   = (UINT8)PciRootBridgeInfo->RootBridge[Index].Bus.Limit;
+
+    DEBUG ((DEBUG_INFO, "PciRootBridgeInfo->RootBridge[Index].Segment %x\n", PciRootBridgeInfo->RootBridge[Index].Segment));
+    DEBUG ((DEBUG_INFO, "UplSegmentInfo->SegmentInfo[Index].BaseAddress; %x\n", UplSegmentInfo->SegmentInfo[Index].BaseAddress));
+    DEBUG ((DEBUG_INFO, "PciRootBridgeInfo->RootBridge[Index].Bus.Base %x\n", PciRootBridgeInfo->RootBridge[Index].Bus.Base));
+    DEBUG ((DEBUG_INFO, "PciRootBridgeInfo->RootBridge[Index].Bus.Limit %x\n", PciRootBridgeInfo->RootBridge[Index].Bus.Limit));
+  }
+
+  Index2 = 0;
+  CopyMem (&mPciSegments[0], &pPciSegments[0], sizeof (PCI_SEGMENT_INFO));
+
+  for (Index = 1; Index <= PciRootBridgeInfo->Count; Index++) {
+    if (mPciSegments[Index2].SegmentNumber == pPciSegments[Index].SegmentNumber) {

```

Comment:
when index2 == 1 (2 SegmentNumber data present in mPciSegments array), we need to compare both mPciSegments[0].SegmentNumber and mPciSegments[1].SegmentNumber, if one of them matched the corresponding StartBus and EndBus should be updated.

So we need nested for loop to support such cases.
You might be able to use NumberOfRootBridges variable for how many SegmentNumber data we have created in mPciSegments array and use "Index2 < NumberOfRootBridges" as nested for loop condition to compare if any other rootbridge sharing the existing Segment number.

Reply:
we need similar logic like below to support variety of RB and Segment combinations (may need fine tuning):

~~~

//
// RBs may share the same Segment number, but mPciSegments should not have duplicate segment data.
// 1. if RB Segment is same with existing one, update StartBus and EndBus of existing one
// 2. if RB Segment is different from all existing ones, add it to the existing Segments data buffer.
// pPciSegments is local temporary data buffer that will be freed later (size depends on numbers of RB)
// mPciSegments is global data that will be updated with the pPciSegments for caller to utilize. (size depends on numbers of different PciSegments)
//
NumberOfExistingSegments = 1;
pPciSegments[0].BaseAddress    = UplSegmentInfo->SegmentInfo[0].BaseAddress;
pPciSegments[0].SegmentNumber  = (UINT16)(PciRootBridgeInfo->RootBridge[0].Segment);
pPciSegments[0].StartBusNumber = (UINT8)PciRootBridgeInfo->RootBridge[0].Bus.Base;
pPciSegments[0].EndBusNumber   = (UINT8)PciRootBridgeInfo->RootBridge[0].Bus.Limit;

for (IndexForRb = 1;  IndexForRb < PciRootBridgeInfo->Count; IndexForRb++) {
  for (IndexForExistingSegments = 0;  IndexForExistingSegments < NumberOfExistingSegments; IndexForExistingSegments++) {
    if (pPciSegments[IndexForExistingSegments].SegmentNumber == PciRootBridgeInfo->RootBridge[IndexForRb].Segment) {
      if (pPciSegments[IndexForExistingSegments].StartBusNumber > PciRootBridgeInfo->RootBridge[IndexForRb].Bus.Base) {
          pPciSegments[IndexForExistingSegments].StartBusNumber = (UINT8) PciRootBridgeInfo->RootBridge[IndexForRb].Bus.Base;
      }

      if (pPciSegments[IndexForExistingSegments].EndBusNumber < PciRootBridgeInfo->RootBridge[IndexForRb].Bus.Limit) {
          pPciSegments[IndexForExistingSegments].EndBusNumber = (UINT8) PciRootBridgeInfo->RootBridge[IndexForRb].Bus.Limit;
      }
      break; //breaking after a match found
    }
  } 
  if (IndexForExistingSegments >= NumberOfExistingSegments) {
    //
    // No match found, add it to segments data buffer
    //
    pPciSegments[NumberOfExistingSegments].BaseAddress    = UplSegmentInfo->SegmentInfo[IndexForRb].BaseAddress;
    pPciSegments[NumberOfExistingSegments].SegmentNumber  = (UINT16)(PciRootBridgeInfo->RootBridge[IndexForRb].Segment);
    pPciSegments[NumberOfExistingSegments].StartBusNumber = (UINT8)PciRootBridgeInfo->RootBridge[IndexForRb].Bus.Base;
    pPciSegments[NumberOfExistingSegments].EndBusNumber   = (UINT8)PciRootBridgeInfo->RootBridge[IndexForRb].Bus.Limit;
    NumberOfExistingSegments++;
  }
}

//
// Prepare data for returning to caller
//
*NumberOfRootBridges = NumberOfExistingSegments;
Size = NumberOfExistingSegments * sizeof (PCI_SEGMENT_INFO);
mPciSegments = (PCI_SEGMENT_INFO *) AllocatePool (Size);
CopyMem (&mPciSegments[0], &pPciSegments[0], Size);
if (pPciSegments != NULL) {
  FreePool (pPciSegments);
}

~~~

Reply:
resolved